### PR TITLE
Add ErrorHandler component

### PR DESCRIPTION
**Pull request by yceruto**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | no
| Fixed tickets | #25905, #26448
| License       | MIT
| Doc PR        | TODO

Mainly for API-based apps that don't require TwigBundle to get the correct exception response according to the request format (aka `_format` attribute).

exception_response

:heavy_check_mark: [RFC7807](https://tools.ietf.org/html/rfc7807) compliant for JSON and XML formats.

---

This introduce a new `ErrorRenderer` service that render a `FlattenException` into a given format:
```php
use Symfony\Component\ErrorHandler\ErrorRenderer\ErrorRenderer;
use Symfony\Component\ErrorHandler\ErrorRenderer\HtmlErrorRenderer;
use Symfony\Component\ErrorHandler\ErrorRenderer\JsonErrorRenderer;

$renderers = [
    new HtmlErrorRenderer(),
    new JsonErrorRenderer(),
    // ...
];
$errorRenderer = new ErrorRenderer($renderers);

return new Response(
    $errorRenderer->render($exception, $request->getRequestFormat()), 
    $exception->getStatusCode(), 
    $exception->getHeaders()
);
```

The built-in error renderers are:

| Format | Class |
| --- | --- |
| html | HtmlErrorRenderer |
| json | JsonErrorRenderer |
| xml, atom | XmlErrorRenderer |
| txt | TxtErrorRenderer |

And you can add your own error renderer by implementing the `ErrorRendererInterface` and tagging it with `error_handler.renderer` in your service definition.

Creating your own error renderer for a built-in format will end up replacing the related built-in error renderer.

Demo: https://github.com/yceruto/error-handler-app ([add custom error renderer](https://github.com/yceruto/error-handler-app/commit/06fc647841e327c05515c0da8e2c789aea808228))

### CODE REVIEW

**Comment by Simperfit** [comment]
This is gonna be really nice !

**Comment by nicolas-grekas** [comment]
Would it make sense no name the component `ErrorHandler`?

**Comment by yceruto** [comment]
> Would it make sense no name the component ErrorHandler?

Does that make me think that we should move `ErrorHandler` and` ExceptionHandler` classes from the Debug component to here, also? Currently, this component only solve the convertion of the exception to a formatted output (html, json, etc) but there is no `register()`/`handle()` method involved at this moment.

**Comment by stof** [comment]
The discussion we had with @fabpot and @nicolas-grekas in Brussels concluded that we should indeed create a new component for error handling. Having the production error handling being a responsibility of a component named debug looks weird, as prod is not in debug mode.

**Comment by stof** [comment]
I don't think having an ErrorException class being responsible both for representing the Exception and for rendering it in multiple formats is the right architecture for the component though. This is not extendable at all, and cannot even be configurable (as you cannot instantiate the rendering layer in a place configuring it).

To me, we need 3 *separate* things in the component:

- the ExceptionHandler/ErrorHandler (moved from Debug)
- the value object representing the actual exception
- a renderer layer knowing how to render the exception in different format (HTML, JSON corresponding to some standard format, XML, etc...)

And then, the Twig-based HTML error page should ideally become a hook into this renderer layer rather than being a totally separate rendering system.

**Comment by yceruto** [comment]
I agree, we should discuss also about BC then, there is a lot of exception classes in Debug component used by these two handlers and we wouldn't want this new component require to `symfony/debug` :)

File: src/Symfony/Component/ErrorHandler/ErrorException.php [hunk 1]
```diff
@@ -0,0 +1,472 @@
...
+                return $this->getHtmlContent($exception);
+        }
+    }
+
+    /**
+     * Gets the HTML content associated with the given exception.
+     *
+     * @return string The content as a string
+     */
+    public function getHtmlBody(FlattenException $exception)
```
Comment (Taluu) [inline src/Symfony/Component/ErrorHandler/ErrorException.php]: isn't it strange that this method (and the one fetching the style) are public but the `getJsonContent`, `getHtmlContent` are private ?

**Comment by yceruto** [comment]
*Update*

 - Component renamed to "ErrorHandler" as suggested by Nicolas (it's a better name IMO too).
 - Almost everthing was moved from the Debug component because interdependence between handlers -> classes/interface, except `Debug` class.

Still working in renderer layer...

**Review by Taluu — COMMENTED** [review]
How about being able to support more formats ? Wouldn't it be better to have some sort of factory so that we can use whatever format we want ?

**Comment by yceruto** [comment]
> How about being able to support more formats ? Wouldn't it be better to have some sort of factory so that we can use whatever format we want ?

Thanks @Taluu for your input, yes I'm working on it.

**Comment by yceruto** [comment]
**Update**

- https://github.com/symfony/symfony/pull/31065/commits/408ecc4695d07b2f0ba620e653bfe8f44eae0ffc Added error renderer layer with interface + DI tag `error_handler.renderer` to add custom format/renderer:
```php
interface ErrorRendererInterface
{
    /**
     * Gets the format of the content.
     *
     * @return string The content format
     */
    public static function getFormat(): string;

    /**
     * Renders an Exception and returns the Response content.
     *
     * @return string The Response content as a string
     */
    public function render(FlattenException $exception): string;
}
```
- https://github.com/symfony/symfony/pull/31065/commits/0ccd7e4cca93c3a4a0aa2f56caf86bce20949477 added FrameworkBundle integration.
- https://github.com/symfony/symfony/pull/31065/commits/9365f95807ccd71c3aa94d58e9641eeec5f9e5bd Updated `ExceptionListener` and `ExceptionController`.

**Comment by yceruto** [comment]
> And then, the Twig-based HTML error page should ideally become a hook into this renderer layer rather than being a totally separate rendering system.

There are many changes (a big diff), most are movements + deprecations, so I prefer to leave this last issue for another PR

**Comment by yceruto** [comment]
The PR app sample is ready: https://github.com/yceruto/error-handler-app ([add custom error renderer](https://github.com/yceruto/error-handler-app/commit/42b65d119a349bccc70d4e45e3472dfa754377c4))

**Comment by yceruto** [comment]
I guess fabbot.io failures related to `src/Symfony/Bridge/PhpUnit/Legacy/SymfonyTestsListenerTrait.php` should be ignored, isn't it?

**Comment by yceruto** [comment]
Tests green! ready on my side.

File: src/Symfony/Component/ErrorHandler/ErrorRenderer/ErrorRenderer.php [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+     * @param \Exception|FlattenException $exception An \Exception or FlattenException instance
+     * @param string                      $format    The request format (html, json, xml, etc.)
+     *
+     * @return string The Response content as a string
+     *
+     * @throws \InvalidArgumentException when the renderer does not exist
+     */
+    public function render($exception, string $format = 'html'): string
+    {
+        if (!$this->container->has($format)) {
```
Comment (kevin-verschaeve) [inline src/Symfony/Component/ErrorHandler/ErrorRenderer/ErrorRenderer.php]: this should not be `'error_handler.renderer.'.$format` ?

Comment (yceruto) [reply src/Symfony/Component/ErrorHandler/ErrorRenderer/ErrorRenderer.php]: It'll be the `ErrorRendererInterface::getFormat()`, ref: https://github.com/symfony/symfony/pull/31065/files#diff-ca872077b071c60c06b91a837070daabR47

Comment (yceruto) [reply src/Symfony/Component/ErrorHandler/ErrorRenderer/ErrorRenderer.php]: BTW, similar approach is done for `FragmentRenderer` feature:
https://github.com/symfony/symfony/blob/238f844c72589020767a0c5928e7b6024c598014/src/Symfony/Component/HttpKernel/DependencyInjection/FragmentRendererPass.php#L57

File: src/Symfony/Component/Debug/CHANGELOG.md [hunk 3]
```diff
@@ -8,6 +8,7 @@ CHANGELOG
 * added `Exception\FlattenException::getAsString` and
 `Exception\FlattenException::getTraceAsString` to increase compatibility to php
 exception objects
+* deprecated the `BufferingLogger`, `ErrorHandler` and `ExceptionHandler` classes
```
[the remark this answers is not in the document]
Comment (yceruto) [reply src/Symfony/Component/Debug/CHANGELOG.md]: Hi Javier, you're completely right. I also forgot to update this line with other movements+deprecations. Thanks!

File: src/Symfony/Component/ErrorHandler/ErrorRenderer/HtmlErrorRenderer.php [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
...25 115.78496551513672,2.7054970115423203 117.42159271240234,2.1007001250982285 C118.79354095458984,1.537783369421959 120.44731903076172,2.0457767099142075 121.32200622558594,3.23083733022213 C121.95732116699219,2.9050118774175644 122.59264373779297,2.5791852325201035 123.22796630859375,2.253336176276207 C123.86669921875,2.5821153968572617 124.50543975830078,2.9108948558568954 125.1441650390625,3.23967407643795 C126.05941009521484,2.154020771384239 127.62747192382812,1.5344576686620712 128.986328125,2.1429056972265244 C130.61741638183594,2.716217741370201 131.50650024414062,4.675290569663048 130.9215545654297,6.2884936183691025 C131.8018341064453,6.78548763692379 132.7589111328125,7.1738648265600204 133.5660400390625,7.780336365103722 C133.60182189941406,11.252970680594444 133.56637573242188,14.726140961050987 133.5631103515625,18.199007019400597 C130.18914794921875,20.431867584586143 126.86984252929688,22.74994657933712 123.44108581542969,24.897907242178917 C122.44406127929688,24.897628769278526 121.5834732055664,23.815067276358604 120.65831756591797,23.37616156041622 C119.76387023925781,22.784828171133995 118.87168884277344,22.19007681310177 117.9798355102539,21.59483526647091 z';
```
[the remark this answers is not in the document]
Comment (yceruto) [reply src/Symfony/Component/ErrorHandler/ErrorRenderer/HtmlErrorRenderer.php]: It's a Github issue, maybe because it's a very long line.

**Comment by yceruto** [comment]
Thanks @javiereguiluz for your review, comments addressed!

**Comment by yceruto** [comment]
Now in production, the HTML error page will look like this (by default):
error-500
Same of https://github.com/symfony/symfony/pull/27699 but without TwigBundle, next https://github.com/symfony/symfony/pull/31065#issuecomment-482902891 we will reuse this one there.

File: src/Symfony/Component/ErrorHandler/ErrorRenderer/TxtErrorRenderer.php [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+    /**
+     * HTML-encodes a string.
+     */
+    private function escapeHtml(string $str): string
+    {
+        return htmlspecialchars($str, ENT_COMPAT | ENT_SUBSTITUTE, $this->charset);
+    }
+
+    private function formatPath(string $path, int $line)
```
Comment (onEXHovia) [inline src/Symfony/Component/ErrorHandler/ErrorRenderer/TxtErrorRenderer.php]: `private function formatPath(string $path, int $line): string`

File: src/Symfony/Component/ErrorHandler/ErrorRenderer/ErrorRenderer.php [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+     * @param ErrorRendererInterface[] $renderers
+     */
+    public function __construct(array $renderers)
+    {
+        foreach ($renderers as $renderer) {
+            $this->addRenderer($renderer);
+        }
+    }
+
+    public function addRenderer(ErrorRendererInterface $renderer): void
```
Comment (onEXHovia) [inline src/Symfony/Component/ErrorHandler/ErrorRenderer/ErrorRenderer.php]: Maybe return `$this`?  `public function addRenderer(ErrorRendererInterface $renderer): self`

**Comment by yceruto** [comment]
Thanks @onEXHovia for your review.

**Comment by yceruto** [comment]
I've splitted the main commit into several topics for easy review :)

**Comment by yceruto** [comment]
Now the Travis failures `dep=high` will be normal until the code is merged into `master|5.0`

This is ready on my side and is very prone to conflicts because the large amount of files touched. Please let me know if I should split this PR for a better review or something else.

friendly ping @fabpot @nicolas-grekas, does this proposal come close to what you had in mind?

**Comment by fabpot** [comment]
Thanks for the ping, now that 4.3 is behind us, I will have a look at this PR closely in the next few days.

File: src/Symfony/Component/Debug/CHANGELOG.md [hunk 7]
```diff
@@ -8,6 +8,14 @@ CHANGELOG
...
 `Exception\FlattenException::getTraceAsString` to increase compatibility to php
 exception objects
+* deprecated the `BufferingLogger`, `DebugClassLoader`, `ErrorHandler` and `ExceptionHandler` classes, 
+  they have been moved to the `ErrorHandler` component
+* deprecated the `FatalErrorHandlerInterface`, `ClassNotFoundFatalErrorHandler`,  
+  `UndefinedFunctionFatalErrorHandler` and `UndefinedMethodFatalErrorHandler` classes,
+  they have been moved to the `ErrorHandler` component
+* deprecated the `ClassNotFoundException`, `FatalErrorException`, `FatalThrowableError`,
+  `FlattenException`, `OutOfMemoryException`, `SilencedErrorContext`, `UndefinedFunctionException`,
+  and `UndefinedMethodException`, they have been moved to the `ErrorHandler` component
```
Comment (derrabus) [inline src/Symfony/Component/Debug/CHANGELOG.md]: The block needs to be moved to 4.4.

File: src/Symfony/Component/Debug/DebugClassLoader.php [hunk 8]
```diff
@@ -11,515 +11,13 @@
...
- * not declare the class.
- *
- * @author Fabien Potencier <fabien@symfony.com>
- * @author Christophe Coevoet <stof@notk.org>
- * @author Nicolas Grekas <p@tchwork.com>
- * @author Guilhem Niot <guilhem.niot@gmail.com>
+ * @deprecated since Symfony 4.4, use Symfony\Component\ErrorHandler\DebugClassLoader instead.
  */
-class DebugClassLoader
+class DebugClassLoader extends BaseDebugClassLoader
```
Comment (derrabus) [inline src/Symfony/Component/Debug/DebugClassLoader.php]: I'm not sure if the `DebugClassLoader` really should be moved. To me, the Debug component seems to be a better fit than the new ErrorHandler component.

File: src/Symfony/Component/ErrorHandler/CHANGELOG.md [hunk 9]
```diff
@@ -0,0 +1,7 @@
+CHANGELOG
+=========
+
+4.3.0
```
Comment (derrabus) [inline src/Symfony/Component/ErrorHandler/CHANGELOG.md]: 4.4.0

File: src/Symfony/Component/Debug/CHANGELOG.md (hunk 7 above)
Comment (yceruto) [reply src/Symfony/Component/Debug/CHANGELOG.md]: Moved

File: src/Symfony/Component/ErrorHandler/CHANGELOG.md (hunk 9 above)
Comment (yceruto) [reply src/Symfony/Component/ErrorHandler/CHANGELOG.md]: Fixed

File: src/Symfony/Component/Debug/DebugClassLoader.php (hunk 8 above)
Comment (yceruto) [reply src/Symfony/Component/Debug/DebugClassLoader.php]: IMHO it fit better in the new component `ErrorHandler`: because it's part of the error mechanism:
> Autoloader checking if the class is really defined in the file found.
> 
> The ClassLoader will wrap all registered autoloaders
> and will throw an exception if a file is found but does
> not declare the class.

**Review by fabpot — CHANGES_REQUESTED** [review]
I've just done a quick review. I have more to come.

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/error_renderer.xml [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+        <service id="error_handler.renderer.xml" class="Symfony\Component\ErrorHandler\ErrorRenderer\XmlErrorRenderer">
+            <tag name="error_handler.renderer" format="atom" />
+            <tag name="error_handler.renderer" />
+            <argument>%kernel.debug%</argument>
+            <argument>%kernel.charset%</argument>
+        </service>
+
+        <service id="error_handler.renderer.txt" class="Symfony\Component\ErrorHandler\ErrorRenderer\TxtErrorRenderer">
+            <tag name="error_handler.renderer" format="rdf" />
```
[the remark this answers is not in the document]
Comment (yceruto) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/error_renderer.xml]: I'm not sure really, but it's supported by TwigBundle.

File: src/Symfony/Component/Debug/DebugClassLoader.php (hunk 8 above)
Comment (yceruto) [reply src/Symfony/Component/Debug/DebugClassLoader.php]: Done.

File: src/Symfony/Component/HttpKernel/EventListener/ExceptionListener.php [hunk 11]
```diff
@@ -151,10 +155,16 @@ protected function duplicateRequest(\Exception $exception, Request $request)
     {
         $attributes = [
             'exception' => $exception = FlattenException::create($exception),
-            '_controller' => $this->controller ?: function () use ($exception) {
-                $handler = new ExceptionHandler($this->debug, $this->charset, $this->fileLinkFormat);
+            '_controller' => $this->controller ?: function () use ($exception, $request) {
+                if (null === $this->errorRenderer) {
+                    $this->errorRenderer = new ErrorRenderer([new HtmlErrorRenderer($this->debug, $this->charset, $this->fileLinkFormat)]);
```
[the remark this answers is not in the document]
Comment (yceruto) [reply src/Symfony/Component/HttpKernel/EventListener/ExceptionListener.php]: Sorry I don't understand what you mean exactly.

`null === $this->errorRenderer` that should not happen frequently, unless you are using the standalone component.

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/error_renderer.xml (hunk 10 above)
[the remark this answers is not in the document]
Comment (yceruto) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/error_renderer.xml]: Removed.

File: src/Symfony/Bundle/WebProfilerBundle/Controller/ExceptionController.php [hunk 12]
```diff
@@ -61,9 +65,11 @@ public function showAction($token)
         $template = $this->getTemplate();
 
         if (!$this->twig->getLoader()->exists($template)) {
-            $handler = new ExceptionHandler($this->debug, $this->twig->getCharset(), $this->fileLinkFormat);
+            if (null === $this->errorRenderer) {
+                $this->errorRenderer = new HtmlErrorRenderer($this->debug, $this->charset, $this->fileLinkFormat);
```
[the remark this answers is not in the document]
Comment (yceruto) [reply src/Symfony/Bundle/WebProfilerBundle/Controller/ExceptionController.php]: Done.

File: src/Symfony/Bundle/WebProfilerBundle/Controller/ExceptionController.php [hunk 13]
```diff
@@ -30,14 +30,18 @@ class ExceptionController
     protected $twig;
     protected $debug;
     protected $profiler;
+    private $charset;
     private $fileLinkFormat;
+    private $errorRenderer;
 
-    public function __construct(Profiler $profiler = null, Environment $twig, bool $debug, FileLinkFormatter $fileLinkFormat = null)
+    public function __construct(Profiler $profiler = null, Environment $twig, bool $debug, FileLinkFormatter $fileLinkFormat = null, string $charset = null, HtmlErrorRenderer $errorRenderer = null)
```
[the remark this answers is not in the document]
Comment (yceruto) [reply src/Symfony/Bundle/WebProfilerBundle/Controller/ExceptionController.php]: Removed.

**Comment by yceruto** [comment]
(Travis dep=high failures expected)

Status: Needs Review

**Review by fabpot — APPROVED** [review]
Great work @yceruto 

I'm going to merge this PR now so that we can tweak it later on.

**Comment by fabpot** [comment]
Thank you @yceruto.

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/error_renderer.xml [hunk 14]
```diff
@@ -0,0 +1,39 @@
+<?xml version="1.0" ?>
+
+<container xmlns="http://symfony.com/schema/dic/services"
+           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+           xsi:schemaLocation="http://symfony.com/schema/dic/services https://symfony.com/schema/dic/services/services-1.0.xsd">
+
+    <services>
+        <defaults public="false" />
```
Comment (Tobion) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/error_renderer.xml]: that seems irrelevant because public=false is the default in sf 4

Comment (yceruto) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/error_renderer.xml]: fixed in https://github.com/symfony/symfony/pull/32227, thanks!

File: src/Symfony/Component/ErrorHandler/ErrorRenderer/ErrorRenderer.php [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+    private $renderers = [];
+
+    /**
+     * @param ErrorRendererInterface[] $renderers
+     */
+    public function __construct(array $renderers)
+    {
+        foreach ($renderers as $renderer) {
+            if (!$renderer instanceof ErrorRendererInterface) {
+                throw new \InvalidArgumentException(sprintf('Error renderer "%s" must implement "%s".', \get_class($renderer), ErrorRendererInterface::class));
```
Comment (Tobion) [inline src/Symfony/Component/ErrorHandler/ErrorRenderer/ErrorRenderer.php]: this seems irrelevant because the type is validated using the typehint in `addRenderer`

File: src/Symfony/Component/ErrorHandler/ErrorRenderer/ErrorRenderer.php [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+     * @throws ErrorRendererNotFoundException if no renderer is found
+     */
+    public function render($exception, string $format = 'html'): string
+    {
+        if (!isset($this->renderers[$format])) {
+            throw new ErrorRendererNotFoundException(sprintf('No error renderer found for format "%s".', $format));
+        }
+
+        if (!$exception instanceof FlattenException) {
+            $exception = FlattenException::create($exception);
```
Comment (Tobion) [inline src/Symfony/Component/ErrorHandler/ErrorRenderer/ErrorRenderer.php]: shouldn't this call `createFromThrowable`?

File: src/Symfony/Component/ErrorHandler/ErrorRenderer/ErrorRenderer.php [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+    {
+        $this->renderers[$format] = $renderer;
+
+        return $this;
+    }
+
+    /**
+     * Renders an Exception and returns the Response content.
+     *
+     * @param \Exception|FlattenException $exception An \Exception or FlattenException instance
```
Comment (Tobion) [inline src/Symfony/Component/ErrorHandler/ErrorRenderer/ErrorRenderer.php]: use `\Throwable` instead of `\Exception`?

Comment (yceruto) [reply src/Symfony/Component/ErrorHandler/ErrorRenderer/ErrorRenderer.php]: see https://github.com/symfony/symfony/pull/32221

File: src/Symfony/Component/ErrorHandler/ErrorRenderer/ErrorRenderer.php (hunk 16 above)
Comment (yceruto) [reply src/Symfony/Component/ErrorHandler/ErrorRenderer/ErrorRenderer.php]: see https://github.com/symfony/symfony/pull/32221

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/error_renderer.xml [hunk 18]
```diff
@@ -0,0 +1,39 @@
...
+            <argument>%debug.file_link_format%</argument>
+        </service>
+
+        <service id="error_handler.renderer.json" class="Symfony\Component\ErrorHandler\ErrorRenderer\JsonErrorRenderer">
+            <tag name="error_handler.renderer" />
+            <argument>%kernel.debug%</argument>
+        </service>
+
+        <service id="error_handler.renderer.xml" class="Symfony\Component\ErrorHandler\ErrorRenderer\XmlErrorRenderer">
+            <tag name="error_handler.renderer" format="atom" />
```
Comment (Tobion) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/error_renderer.xml]: I don't think this works. The final format the renderer is registered by is using the static method of the class which is xml, see https://github.com/symfony/symfony/blob/b92e4ed9d4e480ecfb5dfc0f1dd588b283c7e0a6/src/Symfony/Component/ErrorCatcher/ErrorRenderer/ErrorRenderer.php#L38
So the "atom" format is not available.

Comment (yceruto) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/error_renderer.xml]: See `ErrorCatcherPass` it uses the `Symfony\Component\ErrorCatcher\DependencyInjection\ErrorRenderer` instead.

Comment (Tobion) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/error_renderer.xml]: #32232

**Comment by driesvints** [comment]
@fabpot @yceruto removed public methods like `getContent` and `getStylesheet` on the `Symfony\Component\Debug\ExceptionHandler` class are breaking changes according to SemVer. Can you please revert these removals or introduce this new functionality on 5.0 instead?

These changes will, for example break Lumen: https://github.com/laravel/lumen-framework/issues/935

**Comment by Tobion** [comment]
please open a new issue

**Comment by driesvints** [comment]
@Tobion done: https://github.com/symfony/symfony/issues/32371

File: src/Symfony/Component/ErrorHandler/ErrorRenderer/TxtErrorRenderer.php [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+ */
+class TxtErrorRenderer implements ErrorRendererInterface
+{
+    private $debug;
+    private $charset;
+
+    public function __construct(bool $debug = true, string $charset = null)
+    {
+        $this->debug = $debug;
+        $this->charset = $charset ?: (ini_get('default_charset') ?: 'UTF-8');
```
Comment (vudaltsov) [inline src/Symfony/Component/ErrorHandler/ErrorRenderer/TxtErrorRenderer.php]: @yceruto , the `$charset` property is not used in this class. Consider removing it?

Comment (yceruto) [reply src/Symfony/Component/ErrorHandler/ErrorRenderer/TxtErrorRenderer.php]: Removing it in https://github.com/symfony/symfony/pull/31398/files#diff-ae80b93077d647cfffa8d40755febdb1 Thanks !

**Comment by p803** [comment]
ErrorHandler respecting `Content-Type` is great in API projects! Thanks!
